File: ms_cinema/infrastructure/entry-points/reactive-web/src/main/java/co/com/luisgomez29/api/config/SecurityHeadersConfig.java
package co.com.luisgomez29.api.config;

import jakarta.validation.constraints.NotNull;
import org.springframework.http.HttpHeaders;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ServerWebExchange;
import org.springframework.web.server.WebFilter;
import org.springframework.web.server.WebFilterChain;
import reactor.core.publisher.Mono;

@Component
public class SecurityHeadersConfig implements WebFilter {

    private static final String CSP = "default-src 'none'; img-src 'self'; script-src 'self'; style-src 'self'; " +
            "object-src 'none'; frame-ancestors 'none'; upgrade-insecure-requests";

    @Override
    public @NotNull Mono<Void> filter(ServerWebExchange exchange, @NotNull WebFilterChain chain) {
        HttpHeaders headers = exchange.getResponse().getHeaders();
        headers.set("Content-Security-Policy", CSP);
        headers.set("Strict-Transport-Security", "max-age=31536000; includeSubDomains; preload");
        headers.set("X-Content-Type-Options", "nosniff");
        headers.set("Server", "");
        headers.set("Cache-Control", "no-store");
        headers.set("Pragma", "no-cache");
        headers.set("Referrer-Policy", "strict-origin-when-cross-origin");
        return chain.filter(exchange);
    }
}

File: ms_cinema_kt/infrastructure/entry-points/reactive-web/src/main/kotlin/co/com/luisgomez29/api/config/SecurityHeadersConfig.kt
package co.com.luisgomez29.api.config

import org.springframework.stereotype.Component
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebFilter
import org.springframework.web.server.WebFilterChain
import reactor.core.publisher.Mono

@Component
class SecurityHeadersConfig : WebFilter {
    override fun filter(exchange: ServerWebExchange, chain: WebFilterChain): Mono<Void> {
        val path = exchange.request.path.toString()

        if (path.contains("swagger") || path.contains("api-docs")) {
            return chain.filter(exchange)
        }

        val headers = exchange.response.headers
        headers["Content-Security-Policy"] = CSP
        headers["Strict-Transport-Security"] = "max-age=31536000; includeSubDomains; preload"
        headers["X-Content-Type-Options"] = "nosniff"
        headers.remove("Server")
        headers["Cache-Control"] = "no-store"
        headers["Pragma"] = "no-cache"
        headers["Referrer-Policy"] = "strict-origin-when-cross-origin"
        return chain.filter(exchange)
    }

    companion object {
        private const val CSP = "default-src 'none'; img-src 'self'; script-src 'self'; style-src 'self'; " +
                "object-src 'none'; frame-ancestors 'none'; upgrade-insecure-requests"
    }
}